src/geocel/VolumeSurfaceView.hh

```diff
@@ -0,0 +1,180 @@
+ * (surrounding the entire volume) and optional interface surfaces to an
+ * adjacent volume.
+ */
+class VolumeSurfaceView
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SurfaceParamsRef = NativeCRef<SurfaceParamsData>;
+    //!@}
+
+  public:
+    // Construct from params and pre-step volume ID
+    inline CELER_FUNCTION
+    VolumeSurfaceView(SurfaceParamsRef const& params, VolumeId id);
+
+    // ID of the Volume
+    CELER_FORCEINLINE_FUNCTION VolumeId volume_id() const;
+
+    // ID of the boundary surface for this volume, if any
+    CELER_FORCEINLINE_FUNCTION SurfaceId boundary_id() const;
+
+    // Check if this volume has least one interface surface
+    CELER_FORCEINLINE_FUNCTION bool has_interface() const;
+
+    // Find surface ID for a transition to another volume instances
+    inline CELER_FUNCTION SurfaceId
+    find_interface(VolumeInstanceId pre_id, VolumeInstanceId post_id) const;
+
+  private:
+    SurfaceParamsRef const& params_;
+    VolumeId volume_;
+
+    // HELPER FUNCTIONS
+
+    CELER_FORCEINLINE_FUNCTION VolumeSurfaceRecord const& volume_record() const;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from surface parameters and volume ID.
+ */
+CELER_FUNCTION
+VolumeSurfaceView::VolumeSurfaceView(SurfaceParamsRef const& params,
+                                     VolumeId id)
+    : params_(params), volume_(id)
+{
+    CELER_EXPECT(id < params.volume_surfaces.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the volume ID being viewed.
+ */
+CELER_FUNCTION auto VolumeSurfaceView::volume_id() const -> VolumeId
+{
+    return volume_;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the optional boundary surface ID for this volume.
+
+ * If the result is null,
+ */
+CELER_FUNCTION SurfaceId VolumeSurfaceView::boundary_id() const
+{
+    return this->volume_record().boundary;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Check if this volume has at least one interface surface.
+ */
+CELER_FUNCTION bool VolumeSurfaceView::has_interface() const
+{
+    return !this->volume_record().surface.empty();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find the surface ID for a transition between volume instances.
+ *
+ * This searches for the surface ID associated with a pre->post
+ * volume instance transition.
+ * \todo The current implementation as a linear search, which is unsuitable for
+ * complex detectors such as LHCB's RICH, whose pvRichGrandPMTQuartz has 770
+ * specific interfaces. We should either implenent an \c equal_range function
+ * for searching these sorted arrays, or (better) use a hash lookup for {pre,
+ * post} -> surface.
```

Comment:
```suggestion
 * \todo The current implementation is a linear search, which is unsuitable for
 * complex detectors such as LHCB's RICH, whose pvRichGrandPMTQuartz has 770
 * specific interfaces. We should either implement an \c equal_range function
 * for searching these sorted arrays, or (better) use a hash lookup for {pre,
 * post} -> surface.
```
Since we have `lower_bound` and `upper_bound`, seems like it should be pretty simple to make the search more efficient using the first option for now... any reason to hold off on that?

Reply:
I wanted to get the correct implementation in first, but I'll update this now.